File: apps/ios/ExpCore/Sources/API/TeamMembersApi.swift
import Foundation

public struct UpdateRoleInput: Encodable, Sendable {
    public let memberId: String
    public let role: String

    public init(memberId: String, role: String) {
        self.memberId = memberId
        self.role = role
    }
}

public struct RemoveMemberInput: Encodable, Sendable {
    public let memberId: String

    public init(memberId: String) {
        self.memberId = memberId
    }
}

public final class TeamMembersApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    // The self-service `teamMembers.join` procedure was removed:
    // membership is invite-only.

    public func updateRole(accountId: String, memberId: String, role: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "teamMembers.updateRole", input: UpdateRoleInput(memberId: memberId, role: role))
    }

    public func remove(accountId: String, memberId: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "teamMembers.remove", input: RemoveMemberInput(memberId: memberId))
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/TeamMembersApi.kt
package com.exponential.app.data.api

import com.exponential.app.data.db.TeamMemberEntity
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

@Serializable
data class UpdateRoleInput(val memberId: String, val role: String)

@Serializable
data class UpdateRoleResult(val member: TeamMemberEntity)

@Serializable
data class RemoveMemberInput(val memberId: String)

@Serializable
data class OkResult(val ok: Boolean = false, val success: Boolean = false)

@Singleton
class TeamMembersApi @Inject constructor(private val trpc: TrpcClient) {
    suspend fun updateRole(accountId: String, memberId: String, role: String): TeamMemberEntity =
        trpc.mutation(
            accountId,
            path = "teamMembers.updateRole",
            input = UpdateRoleInput(memberId, role),
            inputSerializer = UpdateRoleInput.serializer(),
            outputSerializer = UpdateRoleResult.serializer(),
        ).member

    suspend fun remove(accountId: String, memberId: String) {
        trpc.mutation(
            accountId,
            path = "teamMembers.remove",
            input = RemoveMemberInput(memberId),
            inputSerializer = RemoveMemberInput.serializer(),
            outputSerializer = OkResult.serializer(),
        )
    }
}
